python/lsst/afw/image/slicing.py

```diff
@@ -1,102 +1,123 @@
+    """
+    defaultOrigin = PARENT
     try:
-        _origin = imageSlice[-1]
-    except TypeError:
-        _origin = None
-
-    if isinstance(_origin, ImageOrigin):
-        origin = _origin
-        imageSlice = imageSlice[0] if len(imageSlice) <= 2 else imageSlice[:-1]
-
-    if isinstance(imageSlice, lsst.geom.Box2I):
-        return imageSlice, origin
-
-    if isinstance(imageSlice, slice) and imageSlice.start is None and imageSlice.stop is None:
-        imageSlice = (Ellipsis, Ellipsis,)
-
-    if not (isinstance(imageSlice, tuple) and len(imageSlice) == 2 and
-            sum([isinstance(_, (slice, type(Ellipsis), int)) for _ in imageSlice]) == 2):
-        raise IndexError(
-            "Images may only be indexed as a 2-D slice not %s" % imageSlice)
-
-    # Because we're going to use slice.indices(...) to construct our ranges, and
-    # python doesn't understand PARENT coordinate systems,  we need
-    # to convert slices specified in PARENT coords to LOCAL
-
-    imageSlice, _imageSlice = [], imageSlice
-    for s, wh, z0 in zip(_imageSlice, img.getDimensions(), img.getXY0()):
-        if origin == LOCAL:
-            z0 = 0                      # ignore image's xy0
-
-        if isinstance(s, slice):
-            if z0 != 0:
-                start = s.start if s.start is None or s.start < 0 else s.start - z0
-                stop = s.stop if s.stop is None or s.stop < 0 else s.stop - z0
-                s = slice(start, stop, s.step)
-        elif isinstance(s, int):
-            if s < 0:
-                s += z0 + wh
-            s = slice(s - z0, s - z0 + 1)
+        if isinstance(sliceArgs[-1], ImageOrigin):
+            # Args are already a tuple that includes the origin.
+            if len(sliceArgs) == 2:
+                return sliceArgs[0], sliceArgs[-1]
+            else:
+                return sliceArgs[:-1], sliceArgs[-1]
         else:
-            s = slice(0, wh)
+            # Args are a tuple that does not include the origin; add it to make origin explicit.
+            return sliceArgs, defaultOrigin
+    except TypeError:  # Arg is a scalar; return it along with the default origin.
+        return sliceArgs, defaultOrigin
 
-        imageSlice.append(s)
 
-    x, y = [_.indices(wh) for _, wh in zip(imageSlice, img.getDimensions())]
-    return lsst.geom.Box2I(lsst.geom.Point2I(x[0], y[0]), lsst.geom.Point2I(x[1] - 1, y[1] - 1)), LOCAL
+def handleNegativeIndex(index, size, origin, default):
+    """Handle negative indices passed to image accessors.
+
+    When negative indices are used in LOCAL coordinates, we interpret them as
+    relative to the upper bounds of the array, as in regular negative indexing
+    in Python.
+
+    When negative indices are used in PARENT coordinates, we interpret them as
+    actual negative pixel values.
+    """
+    if index is None:
+        assert default is not None
+        return default
+    if index < 0 and origin == LOCAL:
```

Comment:
Allowing negative indices for `PARENT` was indeed the old behavior, and changing it was intentional.  I went back and forth a bit on it, but the (self-directed) arguments that convinced me disallowing it is better were:
 - The numbers don't actually make sense; you can interpret the usual (`LOCAL`) negative indices as being a sort of periodic boundary condition, but if an image axes runs from, say, 5 to 11, having "-1" mean "10" just doesn't make sense.
 - (Much more importantly) permitting negative indices only when xy0 is nonnegative could make the same algorithm behave fundamentally different on different inputs.  If a piece of code was written and tested with xy0 >= 0 images in mind, it'd be easy to have a subtle bug that wasn't triggered until someone else tried it with an image with xy0 < 0.  The opposite case is probably more rare, but much more serious: someone could write an algorithm that uses negative indices to mean negative positions, assuming an image that *does* have xy0 < 0.  That code would probably be invalid on images that have xy0 >= 0, but it might silently proceed with incorrect behavior if those the negative indices are incorrectly interpreted as offsets from the end.

Reply:
> The numbers don't actually make sense...

I don't think this is any different than `image[:100, :100]` when `XY0 = (10, 10)`, where `None` is replaced by 10 in the slice.

> permitting negative indices only when xy0 is nonnegative could make the same algorithm behave fundamentally differently on different inputs

I see your point and agree that the old behavior is problematic, so I have a few possible resolutions:

1. We could just always interpret negative indices as subtracted from the column/row size, even if the parent has a negative `XY0`. This will give consistent (and pythonic) behavior but only allow the user to select negative coordinates using a bounding box.

1. We could disable *all* index slicing for images with negative bounding boxes. This might be the best (and most pythonic solution) as any notion of negative indices in python is confusing at best. In this scenario, if a user tried to slice an image using indices or slices (instead of a Box2I or Point2I) then they would get an `IndexError: Images with negative coordinates in their bounding box can only be sliced using a Box2I or Point2I`.

Of course if you don't like either of these options we can just leave things the way that they are, but I feel like for most use cases (are there even any camera geom packages that allow negative coordinates?) using regular python indexing is convenient and desirable.

Reply:
Ok, I think that makes sense. I agree that LOCAL should definitely have pythonic negative slicing, and this is probably the best way to avoid confusion with PARENT slicing.